### DISCUSSION TITLE
FullStack .Net - Dúvida conceitual camadas da aplicação

### DISCUSSION THREAD

**Discussion by rodrigograssi** [balta-io/forum#666] [Tira Dúvidas] [answered]
Minha dúvida é: Tenho minha solução organizada nos seguintes projetos: o Dominio, Dominio.Api, Dominio.Infra, Dominho.App (Blazor).
Quando o projeto é BackEnd (o que estou acostumado), o conhecimento do Domínio para na API, a aplicação não conhece do Domínio, certo?
Quando é full stack a aplicação conhece o Domínio e pode acessá-lo ? 

Como eu já fiz o curso Modelando Domínios Ricos, minhas entidades estão modeladas de formas “ricas” com privates set, tipos complexos,  métodos, validações com o Flunt, etc. Faz sentido eu retornar entidades no retorno dos Handlers?

**Comment by andrebaltieri** [comment, marked as answer]
Bom dia, @rodrigograssi, como vai?

Preciso entender mais a questão... 
Como explico no curso Fullstack, os projetos de Infra e Api, não são legais do App referenciar.... já o domínio, não vejo problema desde que não tenha nada sensível.

Você tem uma visão diferente?

**Reply by rodrigograssi** [reply]
Bom tarde @andrebaltieri, estou bem e você?
Obrigado pela resposta. 
Venho estudando alguns conceitos de arquiteturas de software e na hora de por em prática gera um pouco de confusão.
Minha percepção é que faz sentido o App referenciar o domíno para evitar verbosidade de códigos. 
Na arquitetura hexagonal, ou no clean architecture a aplicação tem acesso ao domíno através dos ports and adpters, usercases, ou Comands no CQRS, e nesses casos eu sempre retorno um ViewObject .Assim exponho somente aquilo que o mundo "externo" deva enxergar.
Sua pergunta me fez refletir bastante, No caso se eu usar a Entidade como resposta, ela também não poderia servir como request? Assim, no caso do curso, não precisaria dos Requests e Reponses (a não ser em casos que especificos). Principalemente em casos somente CRUD, poderia até usar a mesma  pagina do Blazor para criar e editar. Simplificaria a vida :)
Talvez os Requests podem ser uma representação da entidade sem os métodos e propriedades de navegação. 
Sei que não existe uma bala de prata.... mas preciso fazer testes para entender o que mais se aplica aqui.

**Reply by andrebaltieri** [reply]
Bom dia, @rodrigograssi, como vai?
Normal isto acontecer... e sempre que revisitar conceitos terá opiniões diferentes :)

Minha percepção é que faz sentido o App referenciar o domíno para evitar verbosidade de códigos.
- Não digo verbosidade, seria para reaproveitar o código já escrito!

Na arquitetura hexagonal, ou no clean architecture a aplicação tem acesso ao domíno através dos ports and adpters, usercases, ou Comands no CQRS, e nesses casos eu sempre retorno um ViewObject.
- Em todas as arquiteturas isto funciona da mesma forma

Assim exponho somente aquilo que o mundo "externo" deva enxergar.
- Exatamente, o princípio da abstração da OOP

Sua pergunta me fez refletir bastante, No caso se eu usar a Entidade como resposta, ela também não poderia servir como request? 
- Poderia, em um modelo Data Driven ela serve exatamente assim!
- https://balta.io/cursos/criando-apis-data-driven-com-aspnet-core-3-e-ef-core-3

Assim, no caso do curso, não precisaria dos Requests e Reponses (a não ser em casos que especificos). 
- Exato

Principalemente em casos somente CRUD, poderia até usar a mesma pagina do Blazor para criar e editar. Simplificaria a vida :)
- Exatamente

Até que você não está tão confuso assim hahahah 🚀
